crates/rust-client/src/transactions/transaction_script_builder.rs

```
@@ -0,0 +1,201 @@
+use alloc::string::String;
+
+use miden_objects::{
+    accounts::{AccountId, AuthSecretKey},
+    assembly::{AssemblyError, ProgramAst},
+    notes::PartialNote,
+    transaction::TransactionScript,
+    Felt,
+};
+use miden_tx::{
+    auth::TransactionAuthenticator, DataStore, TransactionExecutor, TransactionExecutorError,
+};
+
+use super::prepare_word;
+
+pub struct AccountSpecification {
+    pub account_id: AccountId,
+    pub auth: AuthSecretKey,
+    pub capabilities: AccountCapabilities,
+}
+
+pub enum AccountCapabilities {
+    /// The account exposes procedures of the basic wallet.
+    BasicWallet,
+    /// The account is a fungible faucet and exposes procedures of the basic fungible faucet.
+    BasicFungibleFaucet,
+}
+
+impl AccountCapabilities {
+    /// Returns the script body that sends notes to the recipients.
+    fn send_note_procedure(
+        &self,
+        account_id: AccountId,
+        notes: &[PartialNote],
+    ) -> Result<String, TransactionScriptBuilderError> {
+        let mut body = String::new();
+
+        for partial_note in notes.iter() {
+            if partial_note.metadata().sender() != account_id {
+                return Err(TransactionScriptBuilderError::InvalidSenderAccount(
+                    partial_note.metadata().sender(),
+                ));
+            }
+
+            if partial_note.assets().num_assets() > 1 {
+                return Err(TransactionScriptBuilderError::TooManyAssets(
+                    partial_note.assets().num_assets(),
+                ));
+            }
+
+            let asset = partial_note.assets().iter().next();
+            if let Some(asset) = asset {
```

If a note has no assets, wouldn't this return an invalid script body? I think until we have the ability to send notes without assets, we should make sure there is at least one asset and return an error otherwise.

---

If a note has no assets it wouldn't add anything to the body. The authentication procedure call is later added.

This would be similar to the case where there are no partial output notes (like for the consume note transaction) which currently work.

I'll add a check for empty notes though, until we decide on how we want to treat them.

---

Yeah - I think in this case (until we can handle them properly), we should return an error for notes without assets. Otherwise, we'll create a transaction script which executes but does not output the requested notes.